### PR TITLE
Fix the lua "Build" function getting stuck when prerequisites are missing

### LINKED ISSUE #10617: Lua function "Build" gets stuck when it cannot build units.
I was playing around with .Lua AI for custom missions and I found out that the "Build" function can get the AI stuck while training units if they don't have the necessary prerequisites to train that unit.

Example: AI is trying to train 1tnk, 2tnk and arty (Light Tank, Medium Tank and Artillery respectively), then the player destroys the AI's radar, suddenly the AI stops training Tanks, it won't train Light or Medium tanks either, even though they don't need radar to train, they simply get stuck. I managed to workaround it by performing a constant check if the AI has the prerequisites for each unit it will try to build, but I thought this should be mentioned to the devs. Thanks and sorry if that has been posted before.

### PR DESCRIPTION
**Pull request by abcdefg30**
Fixes #10617.

### CODE REVIEW

**Comment by abcdefg30** [comment]
@Phrohdoh suggested to crash instead of silently failing. I wouldn't really oppose that, just need to go through all missions and add the required checks (to be sure we don't introduce crashes). Any more opinions? (If noone objects, I'll just go for it tomorrow.)

**Comment by Phrohdoh** [comment]
To clarify:

It is my belief that map authors should iterate on their scripts (via playtesting) until all conditions are handled properly instead of passing failures off to the players.

If a prerequisite isn't satisfied the logic should probably queue that up if possible not just silently fail which would probably put the world in a state different from what the script expects.

I'm well aware that we don't have a mechanism right now to queue up a particular structure, place it at (x,y), set the rally point, etc.

**Comment by Phrohdoh** [comment]
Sorry I didn't mean to close this.

**Comment by GraionDilach** [comment]
In an ideal world, @Phrohdoh would be right.

However, map scripting is far, far away from the ideal world. Map scripters lack debuggers and playtesting isn't an option above a certain level. @lovalmidas is a living example of this whom achieved his semi-god map scripting fame half a decade ago by showcasing a mission he couldn't beat himself either.

Regarding the actual use-case - I haven't seen a justification for the explicit crash. In fact, the Build function's documentation clearly says that if it cannot produce the intended actor - regardless of reason - the function should return false. Crashing is definitely not something the scripter should expect.

It would make sense that the Build function should only queue up the actors if all of them have satisfied prerequisites, and only return true in that case with an all-or-nothing mindset.

**Comment by pchote** [comment]
What @GraionDilach suggests makes some sense, however we should be careful: do we really want the function to behave differently if they lose a required prereq one tick before calling the function vs one tick after calling the function?

**Comment by abcdefg30** [comment]
I think a problem here is that `Build` only returns if it could start producing, not if it actually managed to produce something. So, currently, if a prerequisite is lost while building, the `Build` function will have returned true, but the actor won't be produced and the `actionFunc` won't be executed. (In fact, the `actionFunc` will be executed later on, when you produced other actors. E.g. you want to produce 4 actors, where 2 are missing prerequisites. The `actionFunc` will not run after the 2 remaining actors are produced, but after you produce another 2 actors yourself.)

Tl;dr: The entire `Build` function looks rather crap, the more I think about it.

**Comment by lovalmidas** [comment]
(* When I mention "prerequisites" of a unit, this includes the factories from which the unit is supposed to be built from.)

**Possible resolutions:**
- The function performs a pre-check on all prerequisites before production, and only begins if it passes (the above scenario might foil this)
- The function stops producing on invalid prerequisites (no crash), and disposes any further handling (no actionFunc call). **(Similar to current, but no actionFunc call)**
- The function pause on invalid prerequisites, but can continue if the prereq is met later, or stop if a certain time is exceeded. **(similar to RA2)**
- The function skips the actor on invalid prerequisites, and continues to the next.
- The function returns error on invalid prerequisites.
- The function ignores prerequisite checks totally, WW-style. (However, checks for producing factories should still be done)
- Give an extra parameter failureFunc on the function to be called if it encounters invalid prerequisites, or failure in general (e.g. factory exploded).
- A mixture of the above, and give an extra parameter on the Lua function to dictate how such situations should be handled. 

We might want to decide which resolution would be the way to go.

The boolean return will likely have limited usefulness anyway because this function initialises another process (production) over several ticks. Good for pre-checks _only_.

Logging warnings without crashing on encountering invalid prereqs, like what is done in this PR, will be useful to debug existing (and future) missions. Some of the missions (Allied 07 and Allied 10a) gave the Soviets incomplete tech as well (really, WW? No Soviet dome on Allied 10a?).

-------------------
I handle AI team production in my Lua script by maintaining a registry array to queue the actors to be produced for a player, and another for the local AI team. The equivalent of my Build function only appends this list with an actortype entry, and the entry is cleared only after if it is built. So, something along this sequence:

**Pseudo-log**
AITeam1: I want e1, e1, arty.
Registry: e1, e1, arty.
Produce: e1 (Registry -> e1, arty)
Produce: e1 (Registry -> arty)
Prereq check fail: arty (no dome) (Registry remains: arty)
AITeam2: I want e3, e3, 1tnk
Registry: arty, e3, e3, 1tnk
Prereq check fail: arty (no dome) (Registry -> e3, e3, 1tnk, arty)
Produce: e3 (Registry -> e3, 1tnk, arty)
Produce: e3 (Registry -> 1tnk, arty)
Produce: 1tnk (Registry -> arty)
AITeam2 has all members, executing script 1...
Prereq check fail: arty (no dome) (Registry -> e3, e3, 1tnk, arty)
(AI builds dome at some point)
Produce: arty (Registry -> nil)

(Note: I did not use Build() because I was controlling multiple factory per tab production, cost and buildtime modifiers on the Lua)

I follow this particular way of handling: _The function pause on invalid prerequisites, but can continue if the prereq is met later, or stop if a certain time is exceeded_

**Comment by pchote** [comment]
It sounds like what we really need is per-unit callbacks for production complete and production cancelled instead of a single callback when everything is done.

**Comment by abcdefg30** [comment]
That sounds like we should remove the ability to produce multiple units at once (i.e. give the `Build` function an array of actor types), and rather only support building single actors.

**Comment by forcecore** [comment]
But there should be an engine support for building teams.
Doing checkups on the Lua side is painful:
https://github.com/forcecore/yupgi_alert0/blob/master/lua/ai.py : UTIL_BuildTeam()
Remove what we can't build, remove teams that is started to be produced but got invalid, check the produced unit list ...

Lines 868 through 1088 does that but expecting all mappers and AI scripters to do that doesn't look good.

**Comment by abcdefg30** [comment]
This should wait until I added global scripts for all our default mods.

**Comment by pchote** [comment]
This has been inactive for quite some time, so moving to the Future milestone and closing to help clean up the PR queue
